### PR TITLE
Refactor core::stream, build a shared lib

### PR DESCRIPTION
**Pull request by FloopCZ**
~~This is a **heavy WIP**, don't use, don't look, don't complain.~~

Backward incompatible changes:
- `stream::batch` is renamed to `stream::rebatch`.
- There is no `HiPipe::Python` CMake target anymore, Python functionality is integrated into `HiPipe::Core`.
- `stream::buffer` and `stream::rebatch` cannot be followed by any more stream transformers and have to be the last operations in the pipeline. This is for performance, see the documentation for more details.
- After #17, `dataframe` is not a template anymore.

**Right before merging this one, merge #12 #15 #17.**

**DO NOT SQUASH**

A little compile time statistics of one of our larger projects.

Before (`dev` branch):
```
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:12.58
        Maximum resident set size: 8.57 GB
        Major (requiring I/O) page faults: 180
        File system inputs: 62016
        File system outputs: 238528
```

After (`fill-shared-lib` branch):
```
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:30.61
        Maximum resident set size: 6.05 GB
        Major (requiring I/O) page faults: 2
        File system inputs: 4248
        File system outputs: 163944
```

This was not the main purpose of this pull request, but the improvement in compilation time and memory consumption seems significant.

This PR changes the representation of the stream itself - previously, the type of the stream changed after each transformation, now it stays the same - `forward_stream_t` (which is actually an alias to `ranges::any_view<batch_t>`). I was kind of hoping this feature will allow a new syntax for very long pipelines that will push the compilation time even lower, but I have not experimented with that yet. Some other time hopefully.

### CODE REVIEW

File: include/hipipe/core/stream/column.hpp [hunk 1]
```diff
@@ -54,44 +106,90 @@ class column_base {
...
 
-/// Specialization of column_base for non-copy-constructible types.
-template <typename T>
-struct column_base<T, false> : column_base<T, true> {
-    using column_base<T, true>::column_base;
+    template<typename Column>
+    Column& at()
+    {
+        throw_check_extraction_type<Column>();
+        return *columns_.at(std::type_index{typeid(Column)});
```
Comment (petrbel) [inline include/hipipe/core/stream/column.hpp]: why not `columns_->at` ?

File: include/hipipe/core/stream/column.hpp [hunk 2]
```diff
@@ -54,44 +106,90 @@ class column_base {
...
+        static_assert(std::is_constructible_v<Column, Args&&...>,
+          "Cannot cosntruct the given column from the provided arguments.");
+        columns_[std::type_index{typeid(Column)}] =
+          std::make_unique<Column>(std::forward<Args>(args)...);
+    }
+
+    // column check //
+
+    template<typename Column>
+    bool contains() const
```
Comment (petrbel) [inline include/hipipe/core/stream/column.hpp]: so 2020

File: include/hipipe/core/stream/column.hpp (hunk 1 above)
Comment (FloopCZ) [reply include/hipipe/core/stream/column.hpp]: That would be a different thing. `columns_` is not a pointer, it is a `map`. I want to dereference the pointer retrieved by `at()`. Operator `*` has a lower priority than `.`.

**Review by blazekadam — APPROVED** [review]
God bless thy letter!

**Pull request merged**